components/public-api-server/pkg/idp/cache.go

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) 2023 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License.AGPL.txt in the project root for license information.
+
+package idp
+
+import (
+	"context"
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/gitpod-io/gitpod/common-go/log"
+	"github.com/redis/go-redis/v9"
+	"gopkg.in/square/go-jose.v2"
+)
+
+// KeyCache caches public keys to ensure they're returned with the JWKS as long
+// as there are valid tokens out there using those keys.
+//
+// PoC Note: in production this cache would likely be implemted using Redis or the database.
+type KeyCache interface {
+	// Set rotates the current key
+	Set(ctx context.Context, current *rsa.PrivateKey) error
+
+	// Signer produces a new key signer or nil if Set() hasn't been called yet
+	Signer(ctx context.Context) (jose.Signer, error)
+
+	// PublicKeys returns all un-expired public keys as JSON-encoded *jose.JSONWebKeySet.
+	// This function returns the JSON-encoded form directly instead of the *jose.JSONWebKeySet
+	// to allow for persisted JSON implementations of this interface.
+	PublicKeys(ctx context.Context) ([]byte, error)
+}
+
+type inMemoryKey struct {
+	ID      string
+	Created time.Time
+	Key     *rsa.PublicKey
+}
+
+func NewInMemoryCache() *InMemoryCache {
+	return &InMemoryCache{
+		keys: make(map[string]*inMemoryKey),
+	}
+}
```

Comment:
There's already a library for caching that may be better suited - see how the proxy connections are cached in `pkg/proxy`